File: ts-legacy/src/outputs/x360.ts
import { ButtonEvent, OutputAdapter, X360Mapping, X360ButtonName } from '../types';

const ViGEmClient = require('vigemclient');

type X360OutputOptions = {
  offsetMs: number;
  debug: boolean;
};

type DpadState = {
  up: boolean;
  down: boolean;
  left: boolean;
  right: boolean;
};

const BUTTON_NAMES: X360ButtonName[] = [
  'START',
  'BACK',
  'LEFT_THUMB',
  'RIGHT_THUMB',
  'LEFT_SHOULDER',
  'RIGHT_SHOULDER',
  'GUIDE',
  'A',
  'B',
  'X',
  'Y',
];

function schedule(action: () => void, delayMs: number): void {
  if (delayMs > 0) {
    setTimeout(action, delayMs);
  } else {
    action();
  }
}

export function createX360Output(
  mapping: X360Mapping,
  options: X360OutputOptions
): OutputAdapter {
  const client = new ViGEmClient();
  let controller: any;

  try {
    client.connect();
    controller = client.createX360Controller();
    controller.connect();
  } catch (error) {
    console.error('Failed to initialize ViGEm:', error);
    process.exit(1);
  }

  const dpad: DpadState = {
    up: false,
    down: false,
    left: false,
    right: false,
  };

  function updateDpad(): void {
    const horz = dpad.left ? -1 : dpad.right ? 1 : 0;
    const vert = dpad.up ? 1 : dpad.down ? -1 : 0;
    controller.axis.dpadHorz.setValue(horz);
    controller.axis.dpadVert.setValue(vert);
    controller.update();
  }

  function resolveButton(name: X360ButtonName): any {
    const map: Record<X360ButtonName, any> = {
      START: controller.button.START,
      BACK: controller.button.BACK,
      LEFT_THUMB: controller.button.LEFT_THUMB,
      RIGHT_THUMB: controller.button.RIGHT_THUMB,
      LEFT_SHOULDER: controller.button.LEFT_SHOULDER,
      RIGHT_SHOULDER: controller.button.RIGHT_SHOULDER,
      GUIDE: controller.button.GUIDE,
      A: controller.button.A,
      B: controller.button.B,
      X: controller.button.X,
      Y: controller.button.Y,
    };
    return map[name];
  }

  function setButton(name: X360ButtonName, pressed: boolean): void {
    const button = resolveButton(name);
    if (!button || typeof button.setValue !== 'function') {
      return;
    }
    button.setValue(pressed);
    controller.update();
  }

  function setTrigger(trigger: 'left' | 'right', pressed: boolean): void {
    const value = pressed ? 255 : 0;
    if (trigger === 'left') {
      controller.axis.leftTrigger.setValue(value);
    } else {
      controller.axis.rightTrigger.setValue(value);
    }
    controller.update();
  }

  return {
    handleButton(event: ButtonEvent) {
      const entry = mapping.buttons[event.id.toString()];
      if (!entry) {
        return;
      }

      if (options.debug) {
        console.log(
          `[x360] ${event.pressed ? 'press' : 'release'} button ${event.id}`
        );
      }

      if (entry.type === 'trigger') {
        setTrigger(entry.trigger, event.pressed);
        return;
      }

      schedule(() => {
        if (entry.type === 'dpad') {
          dpad[entry.direction] = event.pressed;
          updateDpad();
          return;
        }

        if (entry.type === 'button') {
          if (!BUTTON_NAMES.includes(entry.name)) {
            return;
          }
          setButton(entry.name, event.pressed);
        }
      }, options.offsetMs);
    },
    shutdown() {
      try {
        for (const name of BUTTON_NAMES) {
          const button = resolveButton(name);
          if (button && typeof button.setValue === 'function') {
            button.setValue(false);
          }
        }
        controller.axis.dpadHorz.setValue(0);
        controller.axis.dpadVert.setValue(0);
        controller.axis.leftTrigger.setValue(0);
        controller.axis.rightTrigger.setValue(0);
        controller.update();
        controller.disconnect();
      } catch (error) {
        console.error('Error during x360 shutdown:', error);
      }
    },
  };
}

File: src/outputs/x360.rs
use crate::types::{
    ButtonEvent, DpadDirection, OutputAdapter, TriggerName, X360ButtonEntry, X360ButtonName,
    X360Mapping,
};

use vigem_client::{Client, TargetId, XButtons, XGamepad, Xbox360Wired};

pub struct X360Output {
    target: Xbox360Wired<Client>,
    gamepad: XGamepad,
    mapping: X360Mapping,
    dpad_up: bool,
    dpad_down: bool,
    dpad_left: bool,
    dpad_right: bool,
    offset_ms: u64,
    debug: bool,
}

impl X360Output {
    pub fn new(mapping: X360Mapping, offset_ms: u64, debug: bool) -> Result<Self, String> {
        let client =
            Client::connect().map_err(|e| format!("Failed to connect to ViGEmBus: {:?}", e))?;
        let id = TargetId::XBOX360_WIRED;
        let mut target = Xbox360Wired::new(client, id);
        target
            .plugin()
            .map_err(|e| format!("Failed to plug in virtual controller: {:?}", e))?;
        target
            .wait_ready()
            .map_err(|e| format!("Failed to wait for controller ready: {:?}", e))?;

        Ok(Self {
            target,
            gamepad: XGamepad::default(),
            mapping,
            dpad_up: false,
            dpad_down: false,
            dpad_left: false,
            dpad_right: false,
            offset_ms,
            debug,
        })
    }

    fn resolve_button_flag(name: &X360ButtonName) -> u16 {
        match name {
            X360ButtonName::Start => XButtons::START,
            X360ButtonName::Back => XButtons::BACK,
            X360ButtonName::LeftThumb => XButtons::LTHUMB,
            X360ButtonName::RightThumb => XButtons::RTHUMB,
            X360ButtonName::LeftShoulder => XButtons::LB,
            X360ButtonName::RightShoulder => XButtons::RB,
            X360ButtonName::Guide => XButtons::GUIDE,
            X360ButtonName::A => XButtons::A,
            X360ButtonName::B => XButtons::B,
            X360ButtonName::X => XButtons::X,
            X360ButtonName::Y => XButtons::Y,
        }
    }

    fn set_button(&mut self, name: &X360ButtonName, pressed: bool) {
        let flag = Self::resolve_button_flag(name);
        if pressed {
            self.gamepad.buttons.raw |= flag;
        } else {
            self.gamepad.buttons.raw &= !flag;
        }
        let _ = self.target.update(&self.gamepad);
    }

    fn update_dpad(&mut self) {
        // Horizontal: left = -32768, right = 32767, center = 0
        let horz: i16 = if self.dpad_left {
            -32768
        } else if self.dpad_right {
            32767
        } else {
            0
        };
        // Vertical: up = 32767, down = -32768, center = 0
        let vert: i16 = if self.dpad_up {
            32767
        } else if self.dpad_down {
            -32768
        } else {
            0
        };

        self.gamepad.thumb_lx = horz;
        self.gamepad.thumb_ly = vert;
        let _ = self.target.update(&self.gamepad);
    }

    fn set_trigger(&mut self, trigger: &TriggerName, pressed: bool) {
        let value = if pressed { 255 } else { 0 };
        match trigger {
            TriggerName::Left => self.gamepad.left_trigger = value,
            TriggerName::Right => self.gamepad.right_trigger = value,
        }
        let _ = self.target.update(&self.gamepad);
    }
}

impl OutputAdapter for X360Output {
    fn handle_button(&mut self, event: &ButtonEvent) {
        let entry = match self.mapping.buttons.get(&event.id.to_string()) {
            Some(e) => e.clone(),
            None => return,
        };

        if self.debug {
            let action = if event.pressed { "press" } else { "release" };
            println!("[x360] {} button {} ", action, event.id);
        }

        if self.offset_ms > 0 {
            std::thread::sleep(std::time::Duration::from_millis(self.offset_ms));
        }

        match &entry {
            X360ButtonEntry::Trigger { trigger } => {
                self.set_trigger(trigger, event.pressed);
            }
            X360ButtonEntry::Dpad { direction } => {
                match direction {
                    DpadDirection::Up => self.dpad_up = event.pressed,
                    DpadDirection::Down => self.dpad_down = event.pressed,
                    DpadDirection::Left => self.dpad_left = event.pressed,
                    DpadDirection::Right => self.dpad_right = event.pressed,
                }
                self.update_dpad();
            }
            X360ButtonEntry::Button { name } => {
                self.set_button(name, event.pressed);
            }
        }
    }

    fn shutdown(&mut self) {
        // Reset all state
        self.gamepad = XGamepad::default();
        let _ = self.target.update(&self.gamepad);
        let _ = self.target.unplug();
    }
}
